.github/workflows/ci-app.yml

```diff
@@ -1,28 +1,80 @@
-name: NextJS CI
+name: React CI
 
 on:
   push:
     branches:
       - main
   pull_request:
+    paths:
+      - app/**
+      - .github/workflows/ci-app.yml
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+defaults:
+  run:
+    working-directory: ./app
 
 jobs:
   test:
-    name: pass/fail checks
+    name: Tests and Formatting
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 18
+          cache-dependency-path: ./app/package-lock.json # or yarn.lock
+          cache: npm # or yarn
+      - run: npm ci
+      - run: npm run format-check
+      - run: npm run test
+
+  # Confirms the app still builds successfully
+  check-app-builds:
+    name: Lint, Type check, and Build
     runs-on: ubuntu-latest
 
-    defaults:
-      run:
-        working-directory: ./app
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 18
+          cache-dependency-path: ./app/package-lock.json # or yarn.lock
+          cache: npm # or yarn
+
+      # https://nextjs.org/docs/advanced-features/ci-build-caching
```

Comment:
ℹ️ I think this actually does make a noticeable impact, and likely a more noticeable one as the site increases in size. In the template:

- With caching: ~12 secs
- Without caching: ~50 secs